Keep HOWTO-chrome.md up to date (#506)
fedidcg/FedCM: explorations/HOWTO-chrome.md

### LoginHint

To use the LoginHint API:

> * Enable the experimental feature `FedCmLoginHint` in `chrome://flags`.

Reviewer: Isn't this generally available since 116 - see https://developer.chrome.com/docs/privacy-sandbox/fedcm-updates/

### UserInfo

To use the UserInfo API:

> * Enable the experimental feature `FedCmLoginHint` in `chrome://flags`.

Reviewer: Isn't this generally available since 116 - see https://developer.chrome.com/docs/privacy-sandbox/fedcm-updates/

### RP Context

To use the RP Context API:

> * Enable the experimental feature `FedCmRpContext` in `chrome://flags`.

Reviewer: This is generally available since 116 isn't it?

Reviewer: See https://developer.chrome.com/docs/privacy-sandbox/fedcm-updates/

Author: Good call. Guess I can remove them in this PR.

5. The browser is going load the `signin_url` when the user is signed-out of the IdP.
6. Call `IdentityProvider.close()` when the user is done logging-in to the IdP.

### Error API

> In order to use the Error API:

Reviewer 2: ```suggestion
To use the Error API:
```

* Enable the experimental feature `FedCmError` in `chrome://flags`.
* Provide an `error` in the id assertion endpoint instead of a `token`:
```
{
  "error" : {
     "code" : "access_denied",
     "url" : "https://idp.example/error?type=foo"
  }
}

> ```

Author: Not sure if the indentation is necessary given that the snippet is rendered by markdown. Keeping it as-is to be consistent with the rest of the doc. Feel free to reopen it. Thanks!

> Note that the `error` field in the response including both `code` and `url` is optional. As long as the flag is enabled, the browser will render an error UI when the token request is failed. The `error` field is used to customize the flow when an error happens. The browser will show a customized UI with proper error message if the code is one of "invalid_request", "unauthorized_client", "access_denied", "server_error", and "temporarily_unavailable". If a `url` field is provided, the browser will add an affordance for users to open a new page (e.g. via pop-up window) with that URL to learn more about the error on that page.

Reviewer 3: This is Chrome specific. Please don't say 'the browser', say 'Chrome'. In particular when talking about the customized UI.

> Note that the `error` field in the response including both `code` and `url` is optional. As long as the flag is enabled, the browser will render an error UI when the token request is failed. The `error` field is used to customize the flow when an error happens. The browser will show a customized UI with proper error message if the code is one of "invalid_request", "unauthorized_client", "access_denied", "server_error", and "temporarily_unavailable". If a `url` field is provided, the browser will add an affordance for users to open a new page (e.g. via pop-up window) with that URL to learn more about the error on that page.

Reviewer 4: Maybe wrap this line?

> ### AccountAutoSelectedFlag API

Reviewer 2: ```suggestion
### `AccountAutoSelectedFlag` API
```

> In order to use the AccountAutoSelectedFlag API:

Reviewer 2: ```suggestion
To use the `AccountAutoSelectedFlag` API:
```

* Enable the experimental feature `FedCmAccountAutoSelectedFlag` in `chrome://flags`.

The browser will send a new boolean to represent whether auto re-authentication was triggered such that the account was auto selected by the browser in the flow to both the IdP and the API caller.

For IdP, the browser will include the boolean `is_account_auto_selected` in the request sent to the id assersion endpoint:
```
POST /fedcm_assertion_endpoint HTTP/1.1
Host: idp.example
Origin: https://rp.example/
Content-Type: application/x-www-form-urlencoded
Cookie: 0x23223
Sec-Fetch-Dest: webidentity

> is_account_auto_selected=true&account_id=123&client_id=client1234&nonce=Ct60bD&disclosure_text_shown=true

Reviewer 3: The new one should go at the end?

* Provide an `error` in the id assertion endpoint instead of a `token`:
```
{
  "error" : {
     "code" : "access_denied",
     "url" : "https://idp.example/error?type=foo"
  }
}
```
Note that the `error` field in the response including both `code` and `url` is
optional. As long as the flag is enabled, Chrome will render an error UI when
the token request fails. The `error` field is used to customize the flow when an
error happens. Chrome will show a customized UI with proper error message if the
code is "invalid_request", "unauthorized_client", "access_denied", "server_error",
or "temporarily_unavailable". If a `url` field is provided, Chrome will add an
affordance for users to open a new page (e.g., via pop-up window) with that URL
to learn more about the error on that page.

### `AccountAutoSelectedFlag` API

To use the AccountAutoSelectedFlag API:
* Enable the experimental feature `FedCmAccountAutoSelectedFlag` in `chrome://flags`.

The browser will send a new boolean to represent whether auto re-authentication was triggered such that the account was auto selected by the browser in the flow to both the IdP and the API caller.

> For IdP, the browser will include the boolean `is_account_auto_selected` in the request sent to the id assersion endpoint:

Reviewer 5: ```suggestion
For IdP, the browser will include the boolean `is_account_auto_selected` in the request sent to the id assertion endpoint:
```

```
POST /fedcm_assertion_endpoint HTTP/1.1
Host: idp.example
Origin: https://rp.example/
Content-Type: application/x-www-form-urlencoded
Cookie: 0x23223
Sec-Fetch-Dest: webidentity

account_id=123&client_id=client1234&nonce=Ct60bD&disclosure_text_shown=true&is_account_auto_selected=true
```

> For the API caller, the browser will include the boolean `isAccountAutoSelected` when resolving the promise with the token:

Reviewer 5: ```suggestion
For the Relying Party, the browser will include the boolean `isAccountAutoSelected` when resolving the promise with the token:
```

Thinking this might be more consistent language here, but not certain.

Author: While it's true that API is called on the RP site, the API caller could be the RP itself or the IdP (via SDK) or 3P library that uses FedCM API.

```
{
  "token": "eyJC...J9.eyJzdWTE2...MjM5MDIyfQ.SflV_adQssw....5c",

>   "isAccountAutoSelected": true

Reviewer 5: providing this information seems like it would be unnecessary. What's the motivation for indicating this to the RP/IDP rather than just making this a client side only piece of information? Say for example, that the user doesn't auto select an account could that reveal information about whether or not it's a temp account versus a primary permanent account based on statistical analysis?

Author: Could you elaborate a bit on the attacking vector here? I'm not sure if I understand what you meant by temp account v.s. primary account. The boolean can only be true for "returning" account which means that the user has used that account in this browser client in the past.

For motivation and privacy & security considerations, please see https://github.com/fedidcg/FedCM/issues/497

Reviewer 5: Thanks, I'll have to give a read into this a bit further. I wasn't aware of the linked issue before this, but just came across it. Do you guys have a glitch.me demo of this feature built by chance? I'd like to play around with this feature a bit more to make sure I'm understanding the goals and am able to assess the potential impact correctly.

FWIW, don't block this on my review but my hunch does think this is probably oversharing information. If I come up with a practical attack then it could always be reverted later too.

Author: You could try https://webid-fcm.glitch.me after enabling the flag. The boolean will be displayed in the log section.
Note: in Chrome, after auto re-authen is triggered, it will be suppressed for 10 mins (or until a user explicitly signs in).

Reviewer 3: Hey Kyle, ok with us merging this howto change or do you have any further comment?

the token request fails. The `error` field is used to customize the flow when an
error happens. Chrome will show a customized UI with proper error message if the
code is "invalid_request", "unauthorized_client", "access_denied", "server_error",
or "temporarily_unavailable". If a `url` field is provided and same-site with
the IdP's `configURL`, Chrome will add an affordance for users to open a new
page (e.g., via pop-up window) with that URL to learn more about the error on
that page.

### `IdentityCredentialAutoSelectedFlag` API

> To use the IdentityCredentialAutoSelectedFlag API:

Reviewer 2: ```suggestion
To use the `IdentityCredentialAutoSelectedFlag` API:
```

Author: Thanks! Not sure if this is needed given the context. Will keep it as-is to be consistent with the rest of the file. Please feel free to reopen if you feel strongly about it or if there's a style guide for this.

* Enable the experimental feature `FedCmIdentityCredentialAutoSelectedFlag` 
in `chrome://flags`.

The browser will send a new boolean to represent whether auto re-authentication
was triggered such that the account was auto selected by the browser in the flow
to both the IdP and the API caller.

For IdP, the browser will include `is_identity_credential_auto_selected` in the

> request sent to the id assersion endpoint:

Reviewer 2: ```suggestion
request sent to the ID assertion endpoint:
```

Author: Thanks! Done.
